### PR TITLE
[FLINK-34690] Cast decimal to VARCHAR as primary key in starrocks sink

### PR DESCRIPTION
**Pull request by loserwang1024**
### What's the problem
Flink only have bigint type, if mysql data is unsigned bigint , maybe out of range, so cast to DECIMAL(20, 0).

If starrocks supports unsigned bigint, maybe can transform DECIMAL(20, 0) to unsigned bigint in sink. To be honest, unsigned bigint is not in SQL standard, thus many databases maybe not support it. https://docs.starrocks.io/zh/docs/sql-reference/sql-statements/data-types/data-type-list/

And starrock not support DECIMAL as primary key, thus should be transformed to VARHCAR.

### How to solve
StarRocks is not support Decimal as primary key, so decimal should be cast to  VARHCAR.

### CODE REVIEW

**Comment by loserwang1024** [comment]
@ruanhang1993 @lvyanquan @banmoy , CC, I am not so familiar with StarRocks.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/main/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksUtils.java [hunk 1]
```diff
@@ -297,10 +297,29 @@ public StarRocksColumn.Builder visit(DoubleType doubleType) {
 
         @Override
         public StarRocksColumn.Builder visit(DecimalType decimalType) {
-            builder.setDataType(DECIMAL);
+            // StarRocks is not support Decimal as primary key, so decimal should be cast to INT,
+            // BIGINT, LARGEINT or VARHCAR.
+            if (!isPrimaryKeys) {
```
Comment (fanqiejiang8) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/main/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksUtils.java:302]: 你好，我想问下这个isPrimaryKeys是在哪里获取的

**Review by fanqiejiang8 — COMMENTED** [review]
isPrimaryKeys变量的来源

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/main/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksUtils.java [hunk 2]
```diff
@@ -297,10 +297,29 @@ public StarRocksColumn.Builder visit(DoubleType doubleType) {
 
         @Override
         public StarRocksColumn.Builder visit(DecimalType decimalType) {
-            builder.setDataType(DECIMAL);
+            // StarRocks is not support Decimal as primary key, so decimal should be cast to INT,
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/main/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksUtils.java:300]:
```suggestion
            // StarRocks does not support Decimal as primary key, so decimal should be cast to INT,
```

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/main/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksUtils.java [hunk 3]
```diff
@@ -297,10 +297,29 @@ public StarRocksColumn.Builder visit(DoubleType doubleType) {
...
+            // BIGINT, LARGEINT or VARHCAR.
+            if (!isPrimaryKeys) {
+                builder.setDataType(DECIMAL);
+                builder.setColumnSize(decimalType.getPrecision());
+                builder.setDecimalDigits(decimalType.getScale());
+            } else if (decimalType.getPrecision() < 10 && decimalType.getScale() == 0) {
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/main/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksUtils.java:306]: `getScale() == 0` check was repeated 3 times. Maybe check it first, and fallback to `VARCHAR` if such DecimalType were used?

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/test/java/org/apache/flink/cdc/connectors/starrocks/sink/CdcDataTypeTransformerTest.java [hunk 4]
```diff
@@ -75,6 +76,68 @@ public void testCharTypeForPrimaryKey() {
...
+        StarRocksColumn bigIntLengthColumn = bigIntLengthBuilder.build();
+        assertEquals("primary_key", bigIntLengthColumn.getColumnName());
+        assertEquals(0, bigIntLengthColumn.getOrdinalPosition());
+        assertEquals(StarRocksUtils.BIGINT, bigIntLengthColumn.getDataType());
+        assertTrue(bigIntLengthColumn.isNullable());
+
+        // map to LARGEINT of StarRocks if primary key column with precision < 18 and scale = 0
+        StarRocksColumn.Builder unsignedBigIntLengthBuilder =
+                new StarRocksColumn.Builder().setColumnName("primary_key").setOrdinalPosition(0);
+        new DecimalType(20, 0)
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/test/java/org/apache/flink/cdc/connectors/starrocks/sink/CdcDataTypeTransformerTest.java:106]: Should add checks when `DecimalType` precision goes across specific boundaries like from 9 to 10, 18 to 19, and 37 to 38. Maybe add test cases with these edge cases.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/test/java/org/apache/flink/cdc/connectors/starrocks/sink/CdcDataTypeTransformerTest.java [hunk 5]
```diff
@@ -75,6 +76,68 @@ public void testCharTypeForPrimaryKey() {
...
+        StarRocksColumn.Builder scaleBuilder =
+                new StarRocksColumn.Builder().setColumnName("primary_key").setOrdinalPosition(0);
+        new DecimalType(12, 1)
+                .accept(new StarRocksUtils.CdcDataTypeTransformer(true, scaleBuilder));
+        StarRocksColumn scaleBoundColumn = scaleBuilder.build();
+        assertEquals("primary_key", scaleBoundColumn.getColumnName());
+        assertEquals(0, scaleBoundColumn.getOrdinalPosition());
+        assertEquals(StarRocksUtils.VARCHAR, scaleBoundColumn.getDataType());
+        assertEquals(Integer.valueOf(12), scaleBoundColumn.getColumnSize().orElse(null));
+        assertTrue(unsignedBigIntLengthColumn.isNullable());
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/test/java/org/apache/flink/cdc/connectors/starrocks/sink/CdcDataTypeTransformerTest.java:138]: Minor suggestion: maybe create an ad-hoc function like `checkPrimaryColumnTypeCast(DataType in, DataType expected)` to reduce duplicated codes?

**Review by yuxiqian — CHANGES_REQUESTED** [review]

**Comment by yuxiqian** [comment]
Thanks for @loserwang1024's contribution, left some comments.

**Comment by loserwang1024** [comment]
@yuxiqian , thanks for your advise. 
I think cast decimal to VARCHAR rather than  VARCHAR/INT/BIGINT/LARGEINT is more easy for user to understand, so I just modified it. Could you please help me review this PR again?

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/main/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksUtils.java [hunk 6]
```diff
@@ -129,7 +129,7 @@ public static void toStarRocksDataType(
      * @param fieldType the element type of the RecordData
      * @param fieldPos the element position of the RecordData
      * @param zoneId the time zone used when converting from <code>TIMESTAMP WITH LOCAL TIME ZONE
-     *     </code>
+     *                  </code>
```
[the remark this answers is not in the document]
Comment (loserwang1024) [reply flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/main/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksUtils.java]: no, it seems some unintentional operation.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/main/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksUtils.java [hunk 7]
```diff
@@ -297,10 +297,17 @@ public StarRocksColumn.Builder visit(DoubleType doubleType) {
...
-            builder.setDataType(DECIMAL);
+            // StarRocks does not support Decimal as primary key, so decimal should be cast to
+            // VARCHAR.
+            if (!isPrimaryKeys) {
+                builder.setDataType(DECIMAL);
+                builder.setColumnSize(decimalType.getPrecision());
+                builder.setDecimalDigits(decimalType.getScale());
+            } else {
+                builder.setDataType(VARCHAR);
+                builder.setColumnSize(Math.min(decimalType.getPrecision(), MAX_VARCHAR_SIZE));
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/main/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksUtils.java:308]: For a `DecimalType` with precision `N`, we may need `N + 1` or `N + 2` characters to store it as a string (one for negative sign, and one for decimal point)

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/test/java/org/apache/flink/cdc/connectors/starrocks/sink/CdcDataTypeTransformerTest.java [hunk 8]
```diff
@@ -75,6 +76,35 @@ public void testCharTypeForPrimaryKey() {
...
+                new StarRocksColumn.Builder().setColumnName("primary_key").setOrdinalPosition(1);
+        new DecimalType(20, 1)
+                .notNull()
+                .accept(new StarRocksUtils.CdcDataTypeTransformer(true, primaryKeyBuilder));
+        StarRocksColumn primaryKeyColumn = primaryKeyBuilder.build();
+        assertEquals("primary_key", primaryKeyColumn.getColumnName());
+        assertEquals(1, primaryKeyColumn.getOrdinalPosition());
+        assertEquals(StarRocksUtils.VARCHAR, primaryKeyColumn.getDataType());
+        assertEquals(Integer.valueOf(20), primaryKeyColumn.getColumnSize().orElse(null));
+        assertTrue(!primaryKeyColumn.isNullable());
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/test/java/org/apache/flink/cdc/connectors/starrocks/sink/CdcDataTypeTransformerTest.java:105]: Should we add test cases covering both zero scaled & non-zero scaled decimal types?

**Review by yuxiqian — CHANGES_REQUESTED** [review]
Thanks @loserwang1024's clarification, left two comments here.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/main/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksUtils.java (hunk 7 above)
Comment (loserwang1024) [reply flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/main/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksUtils.java]: Thank you very much for your advice, I didn't take that into consideration

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/test/java/org/apache/flink/cdc/connectors/starrocks/sink/CdcDataTypeTransformerTest.java (hunk 8 above)
Comment (loserwang1024) [reply flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/test/java/org/apache/flink/cdc/connectors/starrocks/sink/CdcDataTypeTransformerTest.java:105]: Just done it.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/test/java/org/apache/flink/cdc/connectors/starrocks/sink/CdcDataTypeTransformerTest.java [hunk 9]
```diff
@@ -75,6 +76,50 @@ public void testCharTypeForPrimaryKey() {
...
+        StarRocksColumn.Builder unsignedBigIntKeyBuilder =
+                new StarRocksColumn.Builder().setColumnName("primary_key").setOrdinalPosition(1);
+        new DecimalType(20, 0)
+                .notNull()
+                .accept(new StarRocksUtils.CdcDataTypeTransformer(true, unsignedBigIntKeyBuilder));
+        StarRocksColumn unsignedBigIntColumn = unsignedBigIntKeyBuilder.build();
+        assertEquals("primary_key", unsignedBigIntColumn.getColumnName());
+        assertEquals(1, unsignedBigIntColumn.getOrdinalPosition());
+        assertEquals(StarRocksUtils.VARCHAR, unsignedBigIntColumn.getDataType());
+        assertEquals(Integer.valueOf(22), unsignedBigIntColumn.getColumnSize().orElse(null));
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/test/java/org/apache/flink/cdc/connectors/starrocks/sink/CdcDataTypeTransformerTest.java:119]: Is `VARCHAR(21)` sufficient here, since scale is 0 and decimal point does not exist?

Comment (loserwang1024) [reply flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/test/java/org/apache/flink/cdc/connectors/starrocks/sink/CdcDataTypeTransformerTest.java]: Done it

**Review by yuxiqian — APPROVED** [review]
LGTM

**Review by PatrickRen — APPROVED** [review]
@loserwang1024 Thanks for the patch! LGTM
